File: KeyRotator.java

import android.content.Context;
import android.content.SharedPreferences;
import androidx.security.crypto.EncryptedSharedPreferences;
import androidx.security.crypto.MasterKeys;
import java.io.IOException;
import java.security.GeneralSecurityException;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class KeyRotator {

    private static final String PREF_FILE_NAME = "byok_key_vault";
    private static final String KEY_SET_KEY = "api_key_set";

    private final SharedPreferences sharedPreferences;
    private final List<String> apiKeys = new ArrayList<>();
    private int currentIndex = 0;

    public KeyRotator(Context context) throws GeneralSecurityException, IOException {
        String masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC);
        this.sharedPreferences = EncryptedSharedPreferences.create(
                PREF_FILE_NAME,
                masterKeyAlias,
                context,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        );
        loadKeys();
    }

    private void loadKeys() {
        Set<String> keySet = sharedPreferences.getStringSet(KEY_SET_KEY, null);
        if (keySet != null) {
            apiKeys.addAll(keySet);
        }
    }

    public void addKey(String apiKey) {
        apiKeys.add(apiKey);
        saveKeys();
    }

    public void removeKey(String apiKey) {
        apiKeys.remove(apiKey);
        saveKeys();
    }

    private void saveKeys() {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putStringSet(KEY_SET_KEY, new java.util.HashSet<>(apiKeys));
        editor.apply();
    }

    public String getNextKey() {
        if (apiKeys.isEmpty()) {
            return null;
        }
        String key = apiKeys.get(currentIndex);
        currentIndex = (currentIndex + 1) % apiKeys.size();
        return key;
    }
}

File: app/src/main/java/com/example/gemini_ai_chat/KeyRotator.kt

package com.example.gemini_ai_chat

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys
import java.io.IOException
import java.security.GeneralSecurityException

class KeyRotator(context: Context) {

    private val sharedPreferences: SharedPreferences
    private val apiKeys = mutableListOf<String>()
    private var currentIndex = 0

    init {
        try {
            val masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)
            sharedPreferences = EncryptedSharedPreferences.create(
                PREF_FILE_NAME,
                masterKeyAlias,
                context,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
        } catch (e: GeneralSecurityException) {
            throw RuntimeException("Failed to create encrypted shared preferences", e)
        } catch (e: IOException) {
            throw RuntimeException("Failed to create encrypted shared preferences", e)
        }
        loadKeys()
    }

    private fun loadKeys() {
        val keySet = sharedPreferences.getStringSet(KEY_SET_KEY, null)
        if (keySet != null) {
            apiKeys.addAll(keySet)
        }
    }

    fun addKey(apiKey: String) {
        apiKeys.add(apiKey)
        saveKeys()
    }

    fun removeKey(apiKey: String) {
        apiKeys.remove(apiKey)
        saveKeys()
    }

    private fun saveKeys() {
        val editor = sharedPreferences.edit()
        editor.putStringSet(KEY_SET_KEY, HashSet(apiKeys))
        editor.apply()
    }

    fun getNextKey(): String? {
        if (apiKeys.isEmpty()) {
            return null
        }
        val key = apiKeys[currentIndex]
        currentIndex = (currentIndex + 1) % apiKeys.size
        return key
    }

    companion object {
        private const val PREF_FILE_NAME = "byok_key_vault"
        private const val KEY_SET_KEY = "api_key_set"
    }
}
